### ISSUE TITLE
[question] Unreal Engine 3 compatibility?

### ISSUE THREAD

**Issue by s4e-tom** [Stormancer/stormancer-sdk-cpp#1]
###### Hello,

I have been interested in Stormancer for quite some time. We are working on an *Unreal Engine 3* game. The game networking is powered by the excellent Unreal Networking Architecture.

However, we also want to talk to the client when he's navigating through the menus (not sure I'm clear enough), as we have a matchmaking system instead of a simple server browser. There's also a messaging system with chatrooms and a friend list. Sadly, UE3 seems to lack something nice and reliable for this sort of task.

Currently, I am writing our matchmaking and chat server entirely in Node.js - it communicates with the client using the [TcpLink](https://udn.epicgames.com/Three/TcpLink.html) class, which has some limitations (and that's not native C++ code - it's an UnrealScript class). It works, but if we find something more robust we might replace it, depending on how much time and effort it would take to do so.

Yesterday I had a look at the README of stormancer-sdk-cpp and saw it can be used in an Unreal Engine 4. Has there been any success in making Unreal Engine 3 and Stormancer work together? How difficult would it be, given that we have access to the engine native C++ code?

**Comment by antlafarge** [comment]
Hello,

I am currently fixing a few issues with  Unreal Engine 4 on the latest the Stormancer c++ client SDK. The fixed code will be released this week. I don't see anything that would prevent you to use Stormancer on Unreal Engine 3, it's plain C++ code. The only restriction is C++11 support for the compiler (we are using lambda in our public APIs). 

The server code is developped in C#6.

A room (scene) system is integrated, and we are actually working on an opensource matchmaking component. We don't provide any friendlist implementation yet, but it would be quite easy to create on your side.

**Comment by s4e-tom** [comment]
Since Visual Studio 2010 is required (sigh) to recompile UE3, I guess it boils down to what parts of C++11 are supported or not supported in VS2010. [According to this page](https://msdn.microsoft.com/en-ca/library/vstudio/hh567368.aspx) VS2010 supports Lambdas in their 1.0 version.

**Comment by antlafarge** [comment]
Yes I see. It may be possible if we compile against the VS10 runtime, and if VS10 is able to compile our headers. That's a lot of ifs...

The good news being that we are very interested in providing an Unreal Engine 3 compatible SDK, It may take some time, we are using many C++11 dependencies. Our best bet would be to compile in C++11 with VS10 runtime (is it possible?), then add a layer of code to hide C++11 features in the DLL.

**Comment by s4e-tom** [comment]
Interesting. If I have some time during the month I'll see if the solution provided in the repo compiles in VS2010.

**Comment by antlafarge** [comment]
Okay, Let me know what problems you encounter.

**Comment by s4e-tom** [comment]
Hello. I apologize for the delay.

I tried to compile the solution in VS2010. Sadly, a lot of included headers (mutex for instance, among many others) are not supported until VS2012...
I also tried with VS2015, which has some retrocompatibility (down to VS2010) as long as you have the compilers installed. Of course it provides the same results.

I said that UE3 will only compile with VS2010. That is actually not true. In the native engine code, there are a few lines of codes that check your version of Visual Studio. In UnVcWin32.h...

```
/**
 * We require at least Visual Studio 2010 w/ SP1 to compile
 */
checkAtCompileTime(_MSC_VER >= 1600 && _MSC_FULL_VER >= 160040219, VisualStudio2010_SP1_Required);


/**
 * Future-proofing the min version check so we keep bumping it whenever we upgrade.
 */ 
#if _MSC_VER > 1600
	#pragma message("Detected compiler newer than Visual Studio 2010, please update min version checking in UnVcWin32.h")
#endif

```

You can either modify this block of code or just comment the first line to get it to compile on newer versions like VS2015. You'll still be using the VC10 compiler though.

If one wants to try integrating Stormancer in UE3, I think you'd have to specify a newer compiler (VC11, VC12 or VC14 depending on what's needed) to the C# build code (in UnrealBuildTool IIRC). I didn't have success with that yet.

**Comment by jmderuty** [comment]
Hello,

Yes, It seems that we are using too many Cpp11 features in the the current cpp SDK to be able to compile with VS10. 

I am not suprised neither that you couldn't make it compile with VS15 with the VS10 compiler. 

The good news is that we are currently working with studios who develop on VS10 (with or without UE3), so we intend to work on official support for VS10 at some point in the near future. 

However, that's sadly not at the top of our dev priorities now, and the task looks quite daunting.
It would probably involve creating wrappers for the types we use, and replace our dependencies to rxcpp/pplx with something else, at least on VS10, or maybe update them to compile with VS10.

I will try some ideas next week, I will definitely keep you informed.

**Comment by s4e-tom** [comment]
A daunting task indeed. Good luck!
